File: apps/ios/Packages/Data/Sources/Data/session/APISession.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Common
import Foundation
import Model
import Network
import OpenAPIRuntime
import Persistence
internal import SeerrAPI

struct APISession: Sendable {
    let profiles: ServerProfileStore
    let transport: ProfileTransport

    func use<Value: Sendable>(
        _ id: ProfileID,
        operation: @Sendable (ServerProfile, Client, AuthenticatedTransport) async throws -> Value
    ) async throws -> Value {
        do {
            guard let profile = try await profiles.profiles().first(where: { $0.id == id }) else {
                throw AppError.notFound
            }
            return try await transport.withProfile(profile) { transport in
                try await operation(
                    profile, Client(serverURL: profile.address.apiBase, transport: transport), transport)
            }
        } catch {
            if Task.isCancelled { throw CancellationError() }
            throw appError(error)
        }
    }
}

func appError(_ error: any Error) -> AppError {
    if let client = error as? ClientError { return appError(client.underlyingError) }
    if let error = error as? AppError { return error }
    if error is DecodingError { return .validation }
    if let url = error as? URLError {
        switch url.code {
        case .notConnectedToInternet, .cannotConnectToHost, .cannotFindHost: return .offline
        case .serverCertificateUntrusted, .serverCertificateHasBadDate, .serverCertificateNotYetValid,
            .serverCertificateHasUnknownRoot, .secureConnectionFailed:
            return .tls
        default: return .network
        }
    }
    return .unknown
}

func responseError(_ code: Int) -> AppError {
    switch code {
    case 401: .auth
    case 403: .permission
    case 404: .notFound
    case 400..<500: .validation
    default: .network
    }
}

File: apps/android/core/data/src/main/kotlin/app/gauja/core/data/session/ApiSession.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.data.session

import app.gauja.core.common.AppError
import app.gauja.core.common.AppException
import app.gauja.core.datastore.profiles.ServerProfileStore
import app.gauja.core.model.servers.ProfileId
import app.gauja.core.model.servers.ServerProfile
import app.gauja.core.network.ProfileTransport
import javax.inject.Inject
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.first
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory

internal class ApiSession
@Inject
constructor(
    private val profiles: ServerProfileStore,
    private val transport: ProfileTransport,
    private val json: Json,
) {
    suspend fun <T> use(id: ProfileId, operation: suspend (ServerProfile, Retrofit) -> T): T =
        safeApi {
            transport.withProfile(
                id,
                {
                    profiles.profiles.first().firstOrNull { it.id == id }
                        ?: throw AppException(AppError.NOT_FOUND)
                },
            ) { profile, client ->
                operation(
                    profile,
                    Retrofit.Builder()
                        .baseUrl(profile.address.apiBase)
                        .client(client)
                        .addConverterFactory(
                            json.asConverterFactory("application/json".toMediaType())
                        )
                        .build(),
                )
            }
        }
}

internal suspend fun <T> safeApi(operation: suspend () -> T): T =
    try {
        operation()
    } catch (error: CancellationException) {
        throw error
    } catch (error: AppException) {
        throw error
    } catch (error: java.io.IOException) {
        throw AppException(networkError(error))
    } catch (_: SerializationException) {
        throw AppException(AppError.VALIDATION)
    } catch (_: IllegalArgumentException) {
        throw AppException(AppError.VALIDATION)
    } catch (_: Exception) {
        throw AppException(AppError.UNKNOWN)
    }

internal fun <T> Response<T>.checked(): T {
    if (isSuccessful) return body() ?: throw AppException(AppError.VALIDATION)
    errorBody()?.close()
    throw AppException(
        when (code()) {
            401 -> AppError.AUTH
            403 -> AppError.PERMISSION
            404 -> AppError.NOT_FOUND
            in 400..499 -> AppError.VALIDATION
            else -> AppError.NETWORK
        }
    )
}

// A failed alternate address can wrap the original TLS rejection (OkHttp fast fallback).
private fun networkError(error: Throwable): AppError =
    when {
        error is javax.net.ssl.SSLException ||
            error.cause?.let { networkError(it) == AppError.TLS } == true ||
            error.suppressed.any { networkError(it) == AppError.TLS } -> AppError.TLS
        error is java.net.UnknownHostException || error is java.net.ConnectException ->
            AppError.OFFLINE
        else -> AppError.NETWORK
    }
